text/2019-11-06-unified-binary.md

```diff
@@ -0,0 +1,149 @@
+workflow changes as a result of the upgrade.
+
+## Motivation
+
+### Simpler maintenance
+
+After this change `tikv` will only build and distribute a single binary, reducing
+the number of binaries we must maintain and test.
+
+For users, this would mean they only needed to distribute one binary, and only
+keep one binary up to date.
+
+### Smaller build graph
+
+Running `cargo +nightly build -Z timings --release` allows for the measurement
+of the build time of a near-distributable build.
+
+Timings measurements showed both `tikv-ctl` and `tikv-server` generated build
+tasks that ran in parallel. Making this change would make these a single job
+which would take slightly longer. While this would result in a minor slowdown in
+compile time, the overall outcome would be less CPU time and less disk writes.
+
+### Reduced learning complexity
+
+Currently the `tikv-server` binary has a fairly simple invocation with no subcommands.
+All options are passed as flags (eg. `tikv-server --log-level trace`).
+
+`tikv-ctl` has a much more complex API service. Consider the subcommands of `tikv-ctl`:
+
+```man
+SUBCOMMANDS:
+    bad-regions           Get all regions with corrupt raft
+    cluster               Print the cluster id
+    compact               Compact a column family in a specified range
+    compact-cluster       Compact the whole cluster in a specified range in one or more column families
+    consistency-check     Force a consistency-check for a specified region
+    diff                  Calculate difference of region keys from different dbs
+    dump-snap-meta        Dump snapshot meta file
+    fail                  Inject failures to TiKV and recovery
+    help                  Prints this message or the help of the given subcommand(s)
+    metrics               Print the metrics
+    modify-tikv-config    Modify tikv config, eg. tikv-ctl --host ip:port modify-tikv-config -m kvdb -n
+                          default.disable_auto_compactions -v true
+    mvcc                  Print the mvcc value
+    print                 Print the raw value
+    raft                  Print a raft log entry
+    random-hex            Generate random bytes with specified length and print as hex
+    raw-scan              Print all raw keys in the range
+    recover-mvcc          Recover mvcc data on one node by deleting corrupted keys
+    recreate-region       Recreate a region with given metadata, but alloc new id for it
+    region-properties     Show region properties
+    scan                  Print the range db range
+    size                  Print region size
+    split-region          Split the region
+    store                 Print the store id
+    tombstone             Set some regions on the node to tombstone by manual
+    unsafe-recover        Unsafely recover the cluster when the majority replicas are failed
+```
+
+Moving `tikv-server` to a `tikv server --flags` and keeping all the `tikv-ctl *`
```

Comment:
we also need to consider compatibility when upgrade

Reply:
Yes, this is why we create aliases :)

Reply:
Luckily tikv-ctl and tikv-server don't have command line args conflict(should check again more carefully).
I agree to unified those binaries.